### PR TITLE
rebase of #16105: test_broken, test_skip

### PR DESCRIPTION
**Pull request by StefanKarpinski**
https://github.com/JuliaLang/julia/pull/16105

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Anyone know what's up with the tests why redirecting STDOUT would cause this to hang?

**Comment by yuyichao** [comment]
The buffer is full and your are not reading it?

**Comment by StefanKarpinski** [comment]
Well, it's not my buffer (I'm just rebasing), but sure, that does seem to be it.

**Comment by StefanKarpinski** [comment]
Also, that code isn't added by #16105, so someone else must have broken this recently.

**Comment by StefanKarpinski** [comment]
Confirmed: `make test-test` hangs on my system on master.

200_s

**Comment by StefanKarpinski** [comment]
Hey, @JeffBezanson, can you take a look at the way this changes test invocation quickly, and see if you think the function call overhead is going to be a performance issue? Otherwise I want to merge this.

**Comment by JeffBezanson** [comment]
Nothing jumps out at me --- is there a specific bit you're concerned about?

File: base/test.jl [hunk 1]
```diff
@@ -165,6 +187,46 @@ Returns a `Pass` `Result` if it does, a `Fail` `Result` if it is
 """
 macro test(ex)
     orig_ex = Expr(:inert,ex)
+    result = get_test_result(ex)
+    :(do_test($result, $orig_ex))
```
Comment (StefanKarpinski) [inline base/test.jl]: @JeffBezanson – this

Comment (JeffBezanson) [inline base/test.jl]: Isn't this the same code the macro generated before?

File: base/test.jl [hunk 2]
```diff
@@ -120,6 +127,24 @@ function Base.show(io::IO, t::Error)
     end
 end
 
+"""
+    Broken
+
+The test condition either has been explicitly skipped or is the expected
```
Comment (tkelman) [inline base/test.jl]: these docstrings need to be revised

File: test/test.jl [hunk 3]
```diff
@@ -1,16 +1,28 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-using Base.Test
```
Comment (tkelman) [inline test/test.jl]: this is necessary if you want to run this file without going through `runtests.jl`

File: base/test.jl [hunk 4]
```diff
@@ -294,13 +368,15 @@ type TestSetException <: Exception
...
+    broken::Int
 end
 
 function Base.show(io::IO, ex::TestSetException)
     print(io, "Some tests did not pass: ")
     print(io, ex.pass,  " passed, ")
     print(io, ex.fail,  " failed, ")
-    print(io, ex.error, " errored.")
+    print(io, ex.error, " errored, ")
+    print(io, ex.broken, " broken.")
```
Comment (tkelman) [inline base/test.jl]: so the number of skips is lumped with the number of broken? they seem slightly separate

File: test/test.jl (hunk 3 above)
Comment (StefanKarpinski) [inline test/test.jl]: Isn't that true for every test file?

File: base/test.jl (hunk 4 above)
Comment (StefanKarpinski) [inline base/test.jl]: True, the whole point of having `test_skip` as a separate thing was for reporting.

**Comment by StefanKarpinski** [comment]
TIL that squash and merge doesn't preserve authorship information correctly :\

File: test/test.jl (hunk 3 above)
Comment (tkelman) [inline test/test.jl]: yes, and anywhere someone cared, they've added this

**Comment by tkelman** [comment]
yeah that's the biggest remaining issue with it

**Comment by StefanKarpinski** [comment]
I made a merge commit from the branch with the same change to preserve @morris25's authorship.

**Comment by tkelman** [comment]
how do you do that with the empty merge commit, for future reference?

**Comment by StefanKarpinski** [comment]
I just did `git merge sk/test_broken` and it created the merge commit.

File: test/test.jl (hunk 3 above)
Comment (StefanKarpinski) [inline test/test.jl]: I don't care for it – it's boilerplate. I would be ok with importing `Base.Test` by default, however.

File: base/test.jl (hunk 2 above)
Comment (StefanKarpinski) [inline base/test.jl]: done: 94a54c19edd368008b7efa32192a3cbd36798b0a.

Comment (tkelman) [inline base/test.jl]: Missing a genstdlib run, and I don't think `test_skip` should be exclusively for intermittent failures. I'll take a pass at it later if I don't forget.

File: test/test.jl (hunk 3 above)
Comment (tkelman) [inline test/test.jl]: aren't we trying to move towards things being more modular and having fewer default exports?

Comment (StefanKarpinski) [inline test/test.jl]: I dunno, man. If you want to, add it back.

Comment (tkelman) [inline test/test.jl]: @iainnz added this one in https://github.com/JuliaLang/julia/commit/dc4f1798fe2fa7429e2551caba529cf0b022150d. given he probably won't notice it's gone I don't feel strongly enough about it to add it back here, but @kshyatt has definitely added several of these around the linalg testing and elsewhere

Comment (kshyatt) [inline test/test.jl]: Yes, I have, because I was tired of waiting 10 minutes for `test linalg/triangular` when I wanted to test `diagonal`.

**Pull request merged**